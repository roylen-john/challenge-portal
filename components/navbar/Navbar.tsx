import React, { ReactElement } from 'react'
import Link from 'next/link'
import { useRouter } from 'next/router'
import { Disclosure, Transition } from '@headlessui/react'
import { MenuIcon, XIcon } from '@heroicons/react/outline'
import { classNames } from '../../utils/utils'
import Logo from '../logo/Logo'
import { useAuth } from '../../context/auth/AuthContext'
import { UrlObject } from 'url'

interface navItem {
  name: string
  href?: UrlObject | string
}

export interface iNavbarProps {
  navItems: navItem[]
  onThemeChange: () => void
}

function Navbar({ navItems }: iNavbarProps): ReactElement {
  const router = useRouter()
  const { user, logout } = useAuth()

  return (
    <Disclosure
      as="nav"
      className="bg-neutralBg fixed w-full border-b border-gray-700"
    >
      {({ open }) => (
        <>
          <div className="mx-auto px-2 sm:px-6 lg:px-8">
            <div className="relative flex items-center justify-between h-16">
              <div className="flex-1 flex items-center justify-start sm:items-stretch">
                <div className="flex-shrink-0 flex items-center" role="banner">
                  <span className="h-5 w-auto text-2xl leading-4 lg:mr-5">
                    üê±‚Äçüíª
                  </span>
                  <div className="hidden lg:block">
                    <Logo />
                  </div>
                </div>
                <div className="hidden sm:block sm:ml-6">
                  <ul className="flex space-x-4">
                    {navItems.map((item) => (
                      <li key={item.name}>
                        <Link href={item.href}>
                          <a
                            className={classNames(
                              router.pathname === item.href ||
                                router.pathname ===
                                  (item.href as UrlObject).pathname
                                ? 'transition duration-200 ease-out bg-primaryBold text-white'
                                : 'transition duration-200 ease-out text-contrastNeutralBgSoft hover:bg-primarySoft hover:text-white',
                              'px-3 py-2 rounded text-sm font-medium'
                            )}
                            aria-current={
                              router.pathname === item.href ? 'page' : undefined
                            }
                            role="link"
                          >
                            {item.name}
                          </a>
                        </Link>
                      </li>
                    ))}
                  </ul>
                </div>
              </div>
              {/* Add profile dropdown here */}
              <div className="absolute inset-y-0 right-0 items-center hidden sm:flex">
                {/* Hidden button for theme change
                <button
                  onClick={onThemeChange}
                  className="bg-primaryBold hover:bg-primarySoft hover:text-contrastPrimaryBgSoft text-white px-3 py-2 rounded text-sm font-medium"
                >
                  Theme Toggle
                </button> */}
                {user && (
                  <span className="text-lg font-bold text-white mr-10">
                    {user.name}
                  </span>
                )}
                <button
                  onClick={logout}
                  className="bg-red-600 hover:bg-red-800 text-white px-3 py-2 rounded text-sm font-medium"
                >
                  Logout
                </button>
              </div>
              <div className="absolute inset-y-0 right-0 flex items-center sm:hidden">
                {/* Mobile menu button*/}
                <Disclosure.Button className="inline-flex items-center justify-center p-2 rounded-md text-contrastNeutralBgSoft hover:bg-neutral hover:text-neutralBg focus:outline-none focus:ring-2 focus:ring-inset focus:ring-primaryBold">
                  <span className="sr-only">Open main menu</span>
                  {open ? (
                    <XIcon className="block h-6 w-6" aria-hidden="true" />
                  ) : (
                    <MenuIcon className="block h-6 w-6" aria-hidden="true" />
                  )}
                </Disclosure.Button>
              </div>
            </div>
          </div>

          <Transition
            enter="transition duration-100 ease-out"
            enterFrom="transform -translate-y-16  opacity-0"
            enterTo="transform translate-y-0 opacity-100"
            leave="transition duration-75 ease-out"
            leaveFrom="transform opacity-100"
            leaveTo="transform opacity-0"
          >
            <Disclosure.Panel className="sm:hidden">
              <ul className="px-2 pt-2 pb-3 space-y-1">
                {navItems.map((item) => (
                  <li key={item.name}>
                    <Link href={item.href}>
                      <a
                        className={classNames(
                          router.pathname === item.href
                            ? 'transition duration-200 ease-out bg-primaryBold text-white'
                            : 'transition duration-200 ease-out text-contrastNeutralBgSoft hover:bg-primarySoft hover:text-white',
                          'block px-3 py-2 rounded text-base font-medium'
                        )}
                        aria-current={
                          router.pathname === item.href ? 'page' : undefined
                        }
                        role="link"
                      >
                        {item.name}
                      </a>
                    </Link>
                  </li>
                ))}
              </ul>
            </Disclosure.Panel>
          </Transition>
        </>
      )}
    </Disclosure>
  )
}

export default Navbar
