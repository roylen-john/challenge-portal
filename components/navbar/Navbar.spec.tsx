import React from 'react'
import { render, within } from '@testing-library/react'
import { screen } from '@testing-library/dom'

import Navbar, { Props } from './Navbar'

const DEFAULT_PROPS: Props = {
  navItems: [],
  onThemeChange: jest.fn(),
}

jest.mock('next/router', () => ({
  useRouter() {
    return {
      route: '/',
      pathname: '',
      query: '',
      asPath: '',
    }
  },
}))

describe('Navbar', () => {
  test('renders correctly', () => {
    const { asFragment } = render(<Navbar {...DEFAULT_PROPS} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('renders the logo', () => {
    const { getByRole } = render(<Navbar {...DEFAULT_PROPS} />)
    const logo = getByRole('banner')
    expect(logo).toBeVisible()
  })

  test('renders the nav links', () => {
    const navItems = [
      { name: 'All Challenges', href: '/' },
      { name: 'New Challenge', href: '/new-challenge' },
    ]

    const { getAllByRole } = render(
      <Navbar {...DEFAULT_PROPS} navItems={navItems} />
    )

    const links = getAllByRole('link')
    expect(links).toHaveLength(navItems.length)

    links.forEach((item, index) => {
      const { getByText } = within(item)
      const { name, href } = navItems[index]
      expect(getByText(name)).toBeVisible()
      expect(screen.getByText(name).closest('a')).toHaveAttribute('href', href)
    })
  })
})
