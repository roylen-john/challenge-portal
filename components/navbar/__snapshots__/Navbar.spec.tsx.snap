// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders correctly 1`] = `
<DocumentFragment>
  <nav
    class="bg-neutralBgSoft w-full fixed"
  >
    <div
      class="mx-auto px-2 sm:px-6 lg:px-8"
    >
      <div
        class="relative flex items-center justify-between h-16"
      >
        <div
          class="flex-1 flex items-center justify-start sm:items-stretch"
        >
          <div
            class="flex-shrink-0 flex items-center"
            role="banner"
          >
            <span
              class="h-5 w-auto text-2xl leading-4 lg:mr-5"
            >
              🐱‍💻
            </span>
            <div
              class="hidden lg:block"
            >
              <svg
                class="fill-current text-primaryBold"
                height="100%"
                style="max-width: 180px;"
                viewBox="0 0 254.644 36.661"
                width="100%"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M14.924 3.9h6.396v32.188h-6.396V21.892H6.396v14.196H0V3.9h6.396v12.688h8.528V3.9zM85.384 0v21.32q.936-1.092 1.898-2.314.962-1.222 1.872-2.522.91-1.3 1.742-2.522a38.83 38.83 0 00.87-1.334 30.449 30.449 0 00.586-.98h7.696q-.936 1.196-2.132 2.6-1.196 1.404-2.47 2.86-1.274 1.456-2.548 2.808-1.274 1.352-2.314 2.496a36.858 36.858 0 011.264 1.279q.65.687 1.345 1.482a62.562 62.562 0 01.199.229q1.508 1.742 2.964 3.614a55.986 55.986 0 012.524 3.51 51.289 51.289 0 01.154.234q.936 1.433 1.582 2.623a23.489 23.489 0 01.368.705h-7.332q-.478-.995-1.184-2.189a44.078 44.078 0 00-.454-.749Q91 31.512 89.856 29.926a62.845 62.845 0 00-2.168-2.837 57.734 57.734 0 00-.146-.179q-.896-1.095-1.685-1.9a17.887 17.887 0 00-.473-.466v11.544h-6.448V1.092L85.384 0zm65.52 36.088h-20.436v-5.304h7.02V9.204h-7.02V3.9h20.436v5.304h-6.968v21.58h6.968v5.304zm102.648-23.66l-.988 5.408q-.884-.26-2.756-.858a12.368 12.368 0 00-2.321-.492 16.175 16.175 0 00-1.891-.106 15.758 15.758 0 00-1.059.033q-1.582.107-2.239.562a1.633 1.633 0 00-.004.003 3.073 3.073 0 00-.39.318q-.468.457-.468.956 0 .572.39 1.014.366.414.96.76a5.555 5.555 0 00.08.046q.645.361 1.495.697a16.758 16.758 0 00.013.005q.858.338 1.846.65 1.404.468 2.912 1.04a12.66 12.66 0 012.182 1.068 11.381 11.381 0 01.548.362q1.222.858 2.002 2.184a5.449 5.449 0 01.63 1.678q.15.728.15 1.572 0 1.508-.572 2.86a5.402 5.402 0 01-1.282 1.823 6.777 6.777 0 01-.616.517q-1.145.853-2.89 1.396a14.993 14.993 0 01-.568.164 14.79 14.79 0 01-1.738.349q-.971.138-2.092.191a30.366 30.366 0 01-1.422.032q-3.172 0-5.408-.728a76.273 76.273 0 01-1.063-.354q-1.298-.444-2.23-.824a23.551 23.551 0 01-.295-.122l.988-5.356q1.82.728 3.874 1.352 2.054.624 4.186.624 2.34 0 3.588-.338a2.506 2.506 0 00.531-.205q.704-.378.717-1.145a1.71 1.71 0 000-.028 1.942 1.942 0 00-.086-.588 1.626 1.626 0 00-.304-.556q-.39-.468-1.092-.806-.702-.338-1.612-.65-.91-.312-1.898-.624-1.404-.416-2.86-.962a12.14 12.14 0 01-2.344-1.171 11.155 11.155 0 01-.308-.207q-1.196-.832-1.976-2.106a5.047 5.047 0 01-.607-1.531q-.173-.761-.173-1.667 0-1.56.598-2.938a6.394 6.394 0 011.382-1.998 7.753 7.753 0 01.49-.446 7.828 7.828 0 011.552-1.001q.772-.383 1.698-.663a12.872 12.872 0 011.878-.409q.968-.14 2.071-.176a23.879 23.879 0 01.783-.013 27.652 27.652 0 012.79.135 22.521 22.521 0 011.656.229 18.552 18.552 0 011.693.378q.885.244 1.641.563a11.026 11.026 0 01.228.099zm-49.816 13.52h-15.86a4.632 4.632 0 00.378 1.899q.487 1.096 1.598 1.871 1.805 1.259 4.565 1.368a13.475 13.475 0 00.531.01q1.924 0 3.666-.416a42.586 42.586 0 001.23-.312q.961-.26 1.708-.52l.884 5.46q-1.664.572-3.536.962t-4.212.39q-3.12 0-5.59-.806a12.124 12.124 0 01-2.803-1.312 10.568 10.568 0 01-1.409-1.08 10.386 10.386 0 01-2.527-3.565 12.129 12.129 0 01-.151-.361 12.608 12.608 0 01-.73-2.706q-.206-1.291-.206-2.754a17.831 17.831 0 01.217-2.851 13.073 13.073 0 01.771-2.817 13.608 13.608 0 011.267-2.432 11.068 11.068 0 011.333-1.65q1.612-1.638 3.692-2.47 2.08-.832 4.264-.832a14.56 14.56 0 013.269.344q3.06.705 4.947 2.854 2.248 2.56 2.696 7.004a23.238 23.238 0 01.112 2.33q0 .624-.026 1.3-.023.605-.067 1.001a6.656 6.656 0 01-.011.091zM73.476 12.376l-1.404 5.148a18.647 18.647 0 00-1.244-.385 24.042 24.042 0 00-1.07-.265 10.169 10.169 0 00-1.03-.171q-.97-.115-2.22-.115a15.884 15.884 0 00-1.61.077q-1.253.128-2.212.469a7.555 7.555 0 00-1.267.58 5.576 5.576 0 00-1.203.928 5.636 5.636 0 00-1.345 2.293 6.543 6.543 0 00-.007.021 9.512 9.512 0 00-.393 2.192 11.179 11.179 0 00-.023.72q0 3.14 1.638 5.024a6.093 6.093 0 00.26.28 5.433 5.433 0 001.88 1.205q1.666.667 4.194.715a20.934 20.934 0 00.4.004q1.508 0 3.146-.208t2.99-.676l.936 5.252a13.73 13.73 0 01-1.267.413q-.67.185-1.439.338a25.771 25.771 0 01-.57.107 20.724 20.724 0 01-1.813.231q-.936.08-1.993.1a37.849 37.849 0 01-.718.007q-3.744 0-6.422-.988-2.678-.988-4.394-2.704-1.716-1.716-2.522-4.056a14.99 14.99 0 01-.778-4.043 17.489 17.489 0 01-.028-1.001 15.034 15.034 0 01.571-4.178 13.93 13.93 0 01.261-.814q.832-2.34 2.548-4.082a11.295 11.295 0 012.717-2.012 14.268 14.268 0 011.625-.744 14.258 14.258 0 012.782-.744q1.607-.27 3.484-.27 2.184 0 3.952.312 1.768.312 3.588 1.04zM176.02 0v35.36a19.582 19.582 0 01-1.565.43q-.774.181-1.655.334a37.993 37.993 0 01-1.096.172q-2.548.364-4.992.364-5.72 0-8.658-3.38a11.359 11.359 0 01-2.355-4.518q-.471-1.715-.562-3.754a22.267 22.267 0 01-.021-.984q0-4.305 1.28-7.303a11.277 11.277 0 011.19-2.135q2.47-3.458 7.306-3.458 1.3 0 2.548.286a9.48 9.48 0 011.204.357 7.232 7.232 0 01.98.449V1.092L176.02 0zM46.592 20.956v14.56q-1.075.239-2.942.539a131.607 131.607 0 01-1.218.189q-2.756.416-6.187.416-2.341 0-4.239-.416a9.931 9.931 0 01-1.812-.574 7.803 7.803 0 01-1.438-.804q-1.352-.962-2.08-2.496a7.038 7.038 0 01-.523-1.592q-.154-.743-.192-1.596a12.904 12.904 0 01-.013-.582 9.059 9.059 0 01.137-1.617q.2-1.105.695-1.971.832-1.456 2.236-2.34a9.403 9.403 0 012.233-1.016 11.373 11.373 0 01.991-.258 17.967 17.967 0 013.55-.388 19.875 19.875 0 01.246-.002q2.277 0 4.077.346a15.163 15.163 0 01.343.07v-.78a4.771 4.771 0 00-.243-1.551 4.1 4.1 0 00-.901-1.491 3.268 3.268 0 00-1.176-.78q-1.087-.442-2.776-.442a23.263 23.263 0 00-3.192.214 21.286 21.286 0 00-.318.046 23.852 23.852 0 00-1.085.191q-.962.194-1.645.433l-.884-5.148a11.08 11.08 0 01.844-.252q1.035-.27 2.536-.502a29.388 29.388 0 014.232-.337 32.197 32.197 0 01.292-.001q2.297 0 4.006.407a10.213 10.213 0 011.012.295 9.426 9.426 0 011.597.724 6.924 6.924 0 011.601 1.252 7.212 7.212 0 011.517 2.502 8.706 8.706 0 01.199.618 13.574 13.574 0 01.42 2.217 17.765 17.765 0 01.1 1.917zm182 0v14.56q-1.075.239-2.942.539a131.607 131.607 0 01-1.218.189q-2.756.416-6.187.416-2.341 0-4.239-.416a9.931 9.931 0 01-1.812-.574 7.803 7.803 0 01-1.438-.804q-1.352-.962-2.08-2.496a7.038 7.038 0 01-.523-1.592q-.154-.743-.192-1.596a12.904 12.904 0 01-.013-.582 9.059 9.059 0 01.137-1.617q.2-1.105.695-1.971.832-1.456 2.236-2.34a9.403 9.403 0 012.233-1.016 11.373 11.373 0 01.991-.258 17.967 17.967 0 013.55-.388 19.875 19.875 0 01.246-.002q2.277 0 4.077.346a15.163 15.163 0 01.343.07v-.78a4.771 4.771 0 00-.243-1.551 4.1 4.1 0 00-.901-1.491 3.268 3.268 0 00-1.176-.78q-1.087-.442-2.776-.442a23.263 23.263 0 00-3.192.214 21.286 21.286 0 00-.318.046 23.852 23.852 0 00-1.085.191q-.962.194-1.645.433l-.884-5.148a11.08 11.08 0 01.844-.252q1.035-.27 2.536-.502a29.388 29.388 0 014.232-.337 32.197 32.197 0 01.292-.001q2.297 0 4.006.407a10.213 10.213 0 011.012.295 9.426 9.426 0 011.597.724 6.924 6.924 0 011.601 1.252 7.212 7.212 0 011.517 2.502 8.706 8.706 0 01.199.618 13.574 13.574 0 01.42 2.217 17.765 17.765 0 01.1 1.917zm-58.968 9.984V17.628a7.409 7.409 0 00-.938-.448 9.37 9.37 0 00-.726-.254 6.683 6.683 0 00-1.846-.285 7.494 7.494 0 00-.13-.001 4.497 4.497 0 00-1.367.196 3.39 3.39 0 00-1.961 1.624 6.616 6.616 0 00-.613 1.516q-.414 1.489-.427 3.599a22.334 22.334 0 000 .137q0 2.555.743 4.382a7.488 7.488 0 00.479.974 4.002 4.002 0 002.889 1.994 5.989 5.989 0 001.037.086q.78 0 1.456-.052a21.064 21.064 0 001.043-.107 23.79 23.79 0 00.361-.049zm18.252-9.412h9.776q0-.988-.26-1.898-.26-.91-.832-1.638a4.185 4.185 0 00-1.333-1.106 4.864 4.864 0 00-.123-.064 3.93 3.93 0 00-.954-.324q-.437-.092-.942-.112a7.245 7.245 0 00-.288-.006q-1.204 0-2.094.387a3.887 3.887 0 00-.064.029q-.91.416-1.508 1.144a5.57 5.57 0 00-.915 1.606 6.25 6.25 0 00-.021.058q-.338.936-.442 1.924zm-147.42 9.828V25.74q-.478-.08-1.108-.144a29.128 29.128 0 00-.4-.038q-.884-.078-1.612-.078-.988 0-1.924.13-.934.13-1.635.441a4.174 4.174 0 00-.003.001 2.865 2.865 0 00-.749.477 2.565 2.565 0 00-.369.407 2.047 2.047 0 00-.318.687q-.074.276-.092.6a3.926 3.926 0 00-.006.221q0 1.664 1.248 2.366 1.035.582 2.57.681a10.121 10.121 0 00.654.021q1.04 0 2.028-.026a19.814 19.814 0 00.704-.03q.348-.022.656-.056a10.653 10.653 0 00.356-.044zm182 0V25.74q-.478-.08-1.108-.144a29.128 29.128 0 00-.4-.038q-.884-.078-1.612-.078-.988 0-1.924.13-.934.13-1.635.441a4.174 4.174 0 00-.003.001 2.865 2.865 0 00-.749.477 2.565 2.565 0 00-.369.407 2.047 2.047 0 00-.318.687q-.074.276-.092.6a3.926 3.926 0 00-.006.221q0 1.664 1.248 2.366 1.035.582 2.57.681a10.121 10.121 0 00.654.021q1.04 0 2.028-.026a19.814 19.814 0 00.704-.03q.348-.022.656-.056a10.653 10.653 0 00.356-.044z"
                  fill-rule="evenodd"
                  font-size="12"
                  stroke-linecap="round"
                  stroke-width=".25mm"
                  vector-effect="non-scaling-stroke"
                />
              </svg>
            </div>
          </div>
          <div
            class="hidden sm:block sm:ml-6"
          >
            <ul
              class="flex space-x-4"
            />
          </div>
        </div>
        <div
          class="absolute inset-y-0 right-0 items-center hidden sm:flex"
        />
        <div
          class="absolute inset-y-0 right-0 flex items-center sm:hidden"
        >
          <button
            class="inline-flex items-center justify-center p-2 rounded-md text-contrastNeutralBgSoft hover:bg-neutral hover:text-neutralBg focus:outline-none focus:ring-2 focus:ring-inset focus:ring-primaryBold"
            id="headlessui-disclosure-button-1"
            type="button"
          >
            <span
              class="sr-only"
            >
              Open main menu
            </span>
            <svg
              aria-hidden="true"
              class="block h-6 w-6"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 6h16M4 12h16M4 18h16"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
